src/process.cr

```diff
@@ -240,25 +242,41 @@ class Process
       end
     end
 
+    reader_pipe, writer_pipe = IO.pipe
+
     @pid = Process.fork_internal(run_hooks: false) do
       begin
-        Process.exec_internal(
-          command,
-          argv,
-          env,
-          clear_env,
-          fork_input || input,
-          fork_output || output,
-          fork_error || error,
-          chdir
-        )
+        reader_pipe.close
+        writer_pipe.close_on_exec = true
+        unless Process.exec_internal(
+            command,
+            argv,
+            env,
+            clear_env,
+            fork_input || input,
+            fork_output || output,
+            fork_error || error,
+            chdir
+          )
+          writer_pipe.write_bytes(Errno.value)
+          writer_pipe.close
+        end
       rescue ex
         ex.inspect_with_backtrace STDERR
       ensure
         LibC._exit 127
       end
     end
 
+    writer_pipe.close
+    bytes = uninitialized UInt8[4]
+    if reader_pipe.read(bytes.to_slice) == 4
```

Comment:
Why not `read_bytes`?

Reply:
Because it raises when the pipe is closed in the forked process.

Reply:
Shouldn't this be `read_fully`?

Reply:
(or something like that, a single `read` is not guaranteed to fill the given slice)

Reply:
Right, it should be `read_fully?` - or better just `read_bytes?`. I doubt there is a practical relevance though, the first four bytes from a pipe are hardly going to be split into multiple reads.

Reply:
Reads (or writes) on a pipe smaller than PIPE_BUF (usually 4096) are guaranteed to no be split. We are safe here.

Reply:
Cool, thanks for the explanation.